esp-wifi/src/wifi/mod.rs

```diff
@@ -1026,18 +1030,21 @@ fn rx_token_consume<R, F>(f: F) -> R
 where
     F: FnOnce(&mut [u8]) -> R,
 {
-    critical_section::with(|cs| {
+    let mut data = critical_section::with(|cs| {
         let mut queue = DATA_QUEUE_RX.borrow_ref_mut(cs);
 
-        let mut data = unwrap!(
+        unwrap!(
             queue.dequeue(),
             "unreachable: transmit()/receive() ensures there is a packet to process"
-        );
-        let buffer = data.as_slice_mut();
-        dump_packet_info(&buffer);
+        )
+    });
+
+    // We handle the received data outside of the critical section because
+    // EspWifiPacketBuffer::drop must not be called in a critical section
+    let buffer = data.as_slice_mut();
+    dump_packet_info(&buffer);
 
-        f(buffer)
-    })
+    f(buffer)
```

Comment:
@bjoernQ do you remember why `f` was called in a critical section? I don't think it's necessary, but I'm also not sure

Reply:
I also don't think its required, `f` is managed by smoltcp. If `f` ends up calling device specific things then those things will have thier own critical sections if needed.